Mathlib/Tactic/FieldSimp/Lemmas.lean

```diff
@@ -0,0 +1,549 @@
+/-
+Copyright (c) 2025 Michael Rothgang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth, Arend Mellendijk, Michael Rothgang
+-/
+import Mathlib.Algebra.BigOperators.Group.List.Basic
+import Mathlib.Algebra.Field.Power
+import Mathlib.Util.Qq
+
+/-! # Lemmas for the field_simp tactic
+
+-/
+
+open List
+
+section zpow'
+variable {α : Type*}
+
+section
+variable [GroupWithZero α]
+
+open Classical in
+/-- This is a variant of integer exponentiation, defined for internal use in the `field_simp` tactic
+implementation. It differs from the usual integer exponentiation in that `0 ^ 0` is `0`, not `1`.
+With this choice, the function `n ↦ a ^ n` is always a homomorphism (`a ^ (n + m) = a ^ n * a ^ m`),
+even if `a` is zero. -/
+noncomputable def zpow' (a : α) (n : ℤ) : α :=
+  if a = 0 ∧ n = 0 then 0 else a ^ n
+
+theorem zpow'_add (a : α) (m n : ℤ) :
+    zpow' a (m + n) = zpow' a m * zpow' a n := by
+  by_cases ha : a = 0
+  · simp [zpow', ha]
+    by_cases hn : n = 0
+    · simp +contextual [hn, zero_zpow]
+    · simp +contextual [hn, zero_zpow]
+  · simp [zpow', ha, zpow_add₀]
+
+theorem zpow'_of_ne_zero_right (a : α) (n : ℤ) (hn : n ≠ 0) : zpow' a n = a ^ n := by
+  simp [zpow', hn]
+
+@[simp]
+lemma zero_zpow' (n : ℤ) : zpow' (0 : α) n = 0 := by
+  simp +contextual only [zpow', true_and, ite_eq_left_iff]
+  intro hn
+  exact zero_zpow n hn
+
+lemma zpow'_eq_zero_iff (a : α) (n : ℤ) : zpow' a n = 0 ↔ a = 0 := by
+  obtain rfl | hn := eq_or_ne n 0
+  · simp [zpow']
+  · simp [zpow', zpow_eq_zero_iff hn]
+    tauto
+
+@[simp]
+lemma one_zpow' (n : ℤ) : zpow' (1 : α) n = 1 := by
+  simp [zpow']
+
+@[simp]
+lemma zpow'_one (a : α) : zpow' a 1 = a := by
+  simp [zpow']
+
+lemma zpow'_zero_eq_div (a : α) : zpow' a 0 = a / a := by
+  simp [zpow']
+  by_cases h : a = 0
+  · simp [h]
+  · simp [h]
+
+lemma zpow'_zero_of_ne_zero {a : α} (ha : a ≠ 0) : zpow' a 0 = 1 := by simp [zpow', ha]
+
+lemma zpow'_neg (a : α) (n : ℤ) : zpow' a (-n) = (zpow' a n)⁻¹ := by
+  simp +contextual [zpow', apply_ite]
+  split_ifs with h
+  · tauto
+  · tauto
+
+lemma zpow'_mul (a : α) (m n : ℤ) : zpow' a (m * n) = zpow' (zpow' a m) n := by
+  by_cases ha : a = 0
+  · simp [ha]
+  by_cases hn : n = 0
+  · rw [hn]
+    simp [zpow', ha, zpow_ne_zero ]
+  by_cases hm : m = 0
+  · rw [hm]
+    simp [zpow', ha]
+  simp [zpow', ha, hm, hn]
+  exact zpow_mul a m n
+
+lemma zpow'_ofNat (a : α) {n : ℕ} (hn : n ≠ 0) : zpow' a n = a ^ n := by
+  rw [zpow'_of_ne_zero_right]
+  · simp
+  exact_mod_cast hn
+
+end
+
+lemma mul_zpow' [CommGroupWithZero α] (n : ℤ) (a b : α) :
+    zpow' (a * b) n = zpow' a n * zpow' b n := by
+  by_cases ha : a = 0
+  · simp [ha]
+  by_cases hb : b = 0
+  · simp [hb]
+  simp [zpow', ha, hb]
+  exact mul_zpow a b n
+
+end zpow'
+
+namespace List
```

Comment:
Similarly I think this should be namespaced. You'll have to give up dot notation locally, but I don't think it reasonable for tactic internals to claim the global `List.prod'` name.